components/req_cpu/src/linux.rs

```diff
@@ -0,0 +1,690 @@
+            }
+
+            if self.collectors.is_empty() {
+                // Block the record thread until a new collector coming.
+                if let Ok(msg) = COLLECTOR_REGISTRATION_CHANNEL.1.recv() {
+                    self.handle_collector_registration_msg(msg);
+                }
+
+                // May wait a long time. So drop out-dated state by resetting.
+                should_reset = true;
+            } else {
+                break;
+            }
+        }
+
+        if should_reset {
+            self.reset();
+        }
+    }
+
+    pub fn handle_thread_registration(&mut self) {
+        while let Ok(ThreadRegistrationMsg {
+            thread_id,
+            shared_ptr,
+        }) = THREAD_REGISTRATION_CHANNEL.1.try_recv()
+        {
+            self.thread_stats.insert(
+                thread_id,
+                ThreadStat {
+                    prev_stat: Stat::default(),
+                    shared_ptr,
+                    prev_tag: None,
+                },
+            );
+        }
+    }
+
+    pub fn record(&mut self) {
+        for (tid, thread_stat) in &mut self.thread_stats {
+            let cur_tag = thread_stat.shared_ptr.take().map(|req_tag| {
+                let tag = req_tag.clone();
+                // Put it back as quickly as possible.
+                assert!(thread_stat.shared_ptr.swap(req_tag).is_none());
+                tag
+            });
+
+            let prev_tag = thread_stat.prev_tag.take();
+
+            if cur_tag.is_some() || prev_tag.is_some() {
+                STAT_TASK_COUNT.inc();
+
+                // If existing current tag, need to store the beginning stat.
+                // If existing previous tag, need to get the end stat to calculate delta.
+                if let Ok(stat) = procinfo::pid::stat_task(*PID, *tid) {
+                    // Accumulate the cpu time for the previous tag.
+                    if let Some(prev_tag) = prev_tag {
+                        let prev_cpu_ticks = (thread_stat.prev_stat.utime as u64)
+                            .wrapping_add(thread_stat.prev_stat.stime as u64);
+                        let current_cpu_ticks = (stat.utime as u64).wrapping_add(stat.stime as u64);
+                        let delta_ms = current_cpu_ticks.wrapping_sub(prev_cpu_ticks) * 1_000
+                            / (*CLK_TCK as u64);
+
+                        *self
+                            .current_window_records
+                            .records
+                            .entry(prev_tag)
+                            .or_insert(0) += delta_ms;
+                    }
+
+                    // Store the beginning stat for the current tag.
+                    if cur_tag.is_some() {
+                        thread_stat.prev_tag = cur_tag;
+                        thread_stat.prev_stat = stat;
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn may_gc(&mut self) -> bool {
```

Comment:
Gc might not be a correct wording. Gc means to clean up some garbage, while we are not cleaning garbages..

Reply:
Dead threads are garbages and this function will clean them.